### ISSUE TITLE
Deprecations (existing inspection): reporting consistency

### ISSUE THREAD

**Issue by kalessil** [kalessil/phpinspectionsea#215]
- [x] implementation of a deprecated method;
- [x] overriding deprecated methods;
- [x] extending deprecated classes/interfaces - **by PS**;
- [x] a method is deprecated, but parents' one is not (methods only, not classes)

**Comment by rentalhost** [comment]
Maybe it should consider too when you deprecate a overrided method, then the parent method should be warned too. I don't know if have some sense of deprecate a child method but not the parent one.

```php
class Parent {
    // Warning: this method was deprecated on Child class.
    public function deprecated() {
    }
}

class Child extends Parent {
    /** @deprecated */
    public function deprecated() {
    }
}
```

**Comment by kalessil** [comment]
@rentalhost : makes sense, adding the case into description

**Comment by funivan** [comment]
If someone decide to deprecate method in the child class. And default one (`Parent::deprecatd`) should be used instead. I think this is not invalid situation to show warning on the parent method. 
We can go deeper. Developers can even deprecate `Child` class (and all its methods) and deny users to use it. What do you think about this situation? How to deprecate `Child` class and all his methods?

**Comment by kalessil** [comment]
@funivan I think in the example from @rentalhost we should report the child and ask developers to deprecate the parent one instead.

@rentalhost, are you agree with me?

**Comment by rentalhost** [comment]
@kalessil (cc @funivan) Exactly, it should only report _method-to-method_ (not _method-to-class_), and just advise that is better deprecate the parent method instead of child method (to make deprecation globals on each child of this parent).

About his first argument, I think that we should think better about it applied to real world, because is a good argument, but I can't think about where I can really do that in some real code - _the need of really deprecate just a child method, instead of parent_.

**Comment by davidmpaz** [comment]
Agree for reporting child deprecation. For the specific case [exposed before](#issuecomment-290071290), I as developer would feel more comfortable if when deprecating some child class method of the package, the inspection warn me to deprecate parent method instead. 

Actually it would be nice not even report on parent that we have deprecated on child but also when in child, report we should move deprecation to parent in order to generalize more. Every client of that base class should be notified then.

Regading thinking, when can be usefull such case: _the need of really deprecate just a child method, instead of parent__, if coming into that point, i think something is not going well, that hierarchy should be revised then, or at least, how its being used. 

It looks to me like Liskov principle is not followed by allowing a child class not fully comply with its parent class contract.
